[tests] Workload tests

radical opened the pull request:
This adds `Aspire.Workload.Tests` which are meant to exercise the `aspire` workload. The individual tests create projects from templates just like a user would, and then run, and validate them.

It uses the existing infrastructure from `Aspire.EndToEnd.Tests` to install sdk+workload for testing.

- Enables tests on Linux, and Windows
- Because we don't have docker support on helix/Windows yet, the docker dependent tests are skipped on windows

## Tests

Two kinds of tests are being added here:

1. Tests that create+build a `aspire-starter` template project, and then some tests interact with that running app to validate it
2. Individual tests exercise the various templates, for example to create `aspire-starter` with [xunit,nunit,mstest] options

- The tests also connect to the dashboard and validate that the expected services show up as expected using `Playwright`.
- Some tests also connect to the webapp in `aspire-starter` template to validate some parts of it.

Fixes https://github.com/dotnet/aspire/issues/47 .

Follow up work is tracked in https://github.com/dotnet/aspire/issues/4222

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/4168)

radical committed:
tests/workloads.proj: rename default target

radical committed:
send-to-helix-inner.proj: improve stopping containers, so it correctly handles 'nothing running' case

radical committed:
ToolCommand: Improve waiting, and cleanup

- also show process run times which is useful in identifying hangs

radical committed:
Add support for Playwright

- also, deploy the chromium browser for helix
  - for this we use a `PlaywrightWrapper` project to invoke to run
    `Main` for `Playwright` and use it to provision the browser.

radical committed:
Add TestUtils.cs

radical committed:
BuildEnvironment: some improvements, and cleanup

radical committed:
Add build support for workload tests

radical committed:
Aspire.Workload.Tests: Add new project with tests

radical committed:
Add Aspire.Shared.WorkloadTesting.targets

radical committed:
Enable workload tests on CI for linux, and windows

radical committed:
ci-helix: cleanup

radical committed:
Add README.md for workload, and endtoend tests

radical committed:
Add dogfood.sh

This can be used to prepare a shell environment for using the
sdk+workload from `artifacts/bin/dotnet-latest`.

`$ source /path-to-aspire-repo/dogfood.sh`

radical committed:
ci: workload-tests - cleanup dotnet-latest/dcp processes

joperezr commented on dogfood.sh:
```diff
@@ -0,0 +1,28 @@
+if [ -z "$ZSH_VERSION" ]; then
```
NIT: should probably have `#!/usr/bin/env bash` at the top line of this script.

joperezr replied:
Also, I don't see in other scripts in the repo us checking if you are using Bash or Zsh, is there a reason why we need that here?

radical replied:
This is meant to be used like `source dogfood.sh`, so that wouldn't be appropriate AFAIU.

radical replied:
`${BASH_SOURCE[0]}` is not supported in `zsh`.

joperezr replied:
I see, I think that having it be a `.sh` will be misleading then if the intent is to do source dogfood.sh. Just curious, why not make it an actual bash script that can be executed? Here is precedent of a very similar use cas in the sdk repo: https://github.com/dotnet/sdk/blob/main/eng/dogfood.sh

radical replied:
If the script is run as `/foo/dogfood.sh` then the changes to the environment will be limited to that script. The documentation in sdk also suggests using `source dogfood.sh`:

https://github.com/dotnet/sdk/blob/9e277c9ce470d35ee4ebe751a5629be8688f5037/documentation/project-docs/developer-guide.md?plain=1#L108

When using `source ..` we are using the current shell to run the commands so the shebang should be not be needed, AFAIU.

radical committed:
fix markdown links

joperezr commented on dogfood.sh:
```diff
@@ -0,0 +1,28 @@
+
+REPO_ROOT=$(cd "${scriptroot}";pwd)
+
+PKG_DIR=${REPO_ROOT}/artifacts/packages/Release/Shipping
```
Should the script instead take configuration parameter?

radical replied:
Sure, we can do that. Auto-detect seemed easier in use.

radical replied:
updated.

joperezr commented on dogfood.sh:
```diff
@@ -0,0 +1,28 @@
+export PATH=$REPO_ROOT/artifacts/bin/dotnet-latest:$PATH
+echo Setting export BUILT_NUGETS_PATH=$PKG_DIR to resolve locally built packages
+export BUILT_NUGETS_PATH=$PKG_DIR
+echo "Use $REPO_ROOT/tests/Aspire.Workload.Tests/data/nuget8.config as your local nuget.config to ensure the packages can be restored"
```
Do we need a dogfood.ps1 for Windows too?

radical replied:
Yes. Coming soon in the next PR 😬

radical replied:
https://github.com/dotnet/aspire/issues/4222

joperezr commented on eng/pipelines/templates/BuildAndTest.yml:
```diff
@@ -21,26 +21,14 @@ parameters:
-      displayName: Build
+  - script: ${{ parameters.buildScript }}
+            -restore -build
+            -pack
```
Since we are now calling pack here always, we should probably start calling sign here too and remove the extra step we have later in this file called "Pack, Sign, and Publish". The original thinking of the split was that packacging and signing took longer so those should happen only after tests have validated things are good, but today they mostly just redo a bunch of stuff that we are already doing here, so might as well just merge the two.

radical replied:
I want to tackle this one in a separate PR. Because for the tests I'm using a more constrained project `tests/workloads.proj` which results in faster installs. In a follow up I will check how to merge these two steps. Or probably be able to re-use parts of it.

joperezr replied:
Sounds good to me, let's just make sure that we have an issue or something to track the de-duplication as that would save a bunch of build time.

radical replied:
https://github.com/dotnet/aspire/issues/4222

joperezr commented on tests/Directory.Build.props:
```diff
@@ -6,6 +6,10 @@
+    <PlaywrightDependenciesDirectory>$(ArtifactsBinDir)playwright-deps</PlaywrightDependenciesDirectory>
+
+    <!-- This is useful for local runs -->
+    <VSTestResultsDirectory>$(ArtifactsTestResultsDir)</VSTestResultsDirectory>
```
I assume they were still saved before, just inside the project directory?

radical replied:
Yes.

joperezr commented on tests/Aspire.Workload.Tests/data/nuget8.config:
```diff
@@ -0,0 +1,17 @@
+    <!-- TEST_RESTORE_SOURCES_INSERTION_LINE -->
+    <add key="built-local" value="%BUILT_NUGETS_PATH%" />
+    <add key="dotnet8" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet8/nuget/v3/index.json" />
+    <add key="nuget.org"  value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
```
I may be wrong, but we may get flagged in compliance for having nuget.org as a feed (even when just used for testing.)

eerhardt replied:
Why do we need nuget.org? We should have everything we need in our normal nuget.config (plus the local path)

radical replied:
Hm I'll try without that for the tests. I found it useful when dogfooding the sdk+workload.

radical replied:
```
ServiceDefaults.csproj : error NU1101:

Unable to find package OpenTelemetry.Exporter.OpenTelemetryProtocol. No packages exist with this id in source(s)
Unable to find package OpenTelemetry.Extensions.Hosting. No packages exist with this id in source(s)
Unable to find package OpenTelemetry.Instrumentation.AspNetCore. No packages exist with this id in source(s)
Unable to find package OpenTelemetry.Instrumentation.Http. No packages exist with this id in source(s)
Unable to find package OpenTelemetry.Instrumentation.Runtime. No packages exist with this id in source(s)

Unable to find package Microsoft.Extensions.Http with version (>= 8.0.0)
Unable to find package Microsoft.Extensions.Configuration.Abstractions with version (>= 8.0.0)
Unable to find package Microsoft.Extensions.DependencyInjection.Abstractions with version (>= 8.0.1)
Unable to find package Microsoft.Extensions.Logging.Abstractions with version (>= 8.0.1)
Unable to find package Microsoft.Extensions.Options with version (>= 8.0.2)
Unable to find package Microsoft.Extensions.Primitives with version (>= 8.0.0)
```

radical replied:
```
Unable to find package Microsoft.Extensions.Http with version (>= 8.0.0)
  - Found 1459 version(s) in dotnet8 [ Nearest version: 8.0.0-rtm.23523.3 ]
  - Found 0 version(s) in /Users/ankj/dev/aspire/artifacts/bin/dotnet-latest/library-packs
Unable to find package Microsoft.Extensions.Configuration.Abstractions with version (>= 8.0.0)
  - Found 1459 version(s) in dotnet8 [ Nearest version: 8.0.0-rtm.23523.3 ]
  - Found 0 version(s) in /Users/ankj/dev/aspire/artifacts/bin/dotnet-latest/library-packs
Unable to find package Microsoft.Extensions.DependencyInjection.Abstractions with version (>= 8.0.1)
  - Found 1459 version(s) in dotnet8 [ Nearest version: 8.0.0-rtm.23523.3 ]
  - Found 0 version(s) in /Users/ankj/dev/aspire/artifacts/bin/dotnet-latest/library-packs
Unable to find package Microsoft.Extensions.Logging.Abstractions with version (>= 8.0.1)
  - Found 1459 version(s) in dotnet8 [ Nearest version: 8.0.0-rtm.23523.3 ]
  - Found 0 version(s) in /Users/ankj/dev/aspire/artifacts/bin/dotnet-latest/library-packs
Unable to find package Microsoft.Extensions.Options with version (>= 8.0.2)
  - Found 1459 version(s) in dotnet8 [ Nearest version: 8.0.0-rtm.23523.3 ]
  - Found 0 version(s) in /Users/ankj/dev/aspire/artifacts/bin/dotnet-latest/library-packs
Unable to find package Microsoft.Extensions.Primitives with version (>= 8.0.0)
  - Found 1459 version(s) in dotnet8 [ Nearest version: 8.0.0-rtm.23523.3 ]
  - Found 0 version(s) in /Users/ankj/dev/aspire/artifacts/bin/dotnet-latest/library-packs
```

eerhardt replied:
Why don't you have a `<add key="dotnet-public" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-public/nuget/v3/index.json" />` entry?

radical replied:
Fixed.

joperezr reviewed (commented):
I think it would be really great if we could add some docs somewhere that explain with detail how are people supposed to run these tests locally. Same goes for the rest of the infra and docker tests that we have added in the past, I don't believe I've seen those docs anywhere. I'm fine if we do that as a follow up change but just let's make sure we do get to it.

radical commented:
> I think it would be really great if we could add some docs somewhere that explain with detail how are people supposed to run these tests locally. Same goes for the rest of the infra and docker tests that we have added in the past, I don't believe I've seen those docs anywhere. I'm fine if we do that as a follow up change but just let's make sure we do get to it.

See `README.md` in `tests/Aspire.Workload.Tests`, and `tests/Aspire.EndToEnd.Tests`.

joperezr commented on tests/Aspire.Workload.Tests/Aspire.Workload.Tests.csproj:
```diff
@@ -0,0 +1,59 @@
+    </Exec>
+
+    <PropertyGroup>
+      <_Regex>^ *(Aspire.Workload.Tests[^\($]+)</_Regex>
```
NIT: `^\s*(Aspire\.Workload\.Tests[^\($]+)`

joperezr commented on tests/Aspire.Workload.Tests/Directory.Build.props:
```diff
@@ -0,0 +1,7 @@
+<Project>
+  <PropertyGroup>
+    <IsWorkloadTestProject>true</IsWorkloadTestProject>
```
Small NIT: it is sometimes helpful to add comments above properties like this to just explain what this will do. For instance, this will import workload testing props and targets.

joperezr commented:
> > I think it would be really great if we could add some docs somewhere that explain with detail how are people supposed to run these tests locally. Same goes for the rest of the infra and docker tests that we have added in the past, I don't believe I've seen those docs anywhere. I'm fine if we do that as a follow up change but just let's make sure we do get to it.
> 
> See `README.md` in `tests/Aspire.Workload.Tests`, and `tests/Aspire.EndToEnd.Tests`.

Oh, missed that! Would be great to link those somewhere from our Contributing docs. Essentially, imagine you are a brand new contributor to the repo and want to contribute a change and want to run all tests locally, but don't know where to start. Having that top level documentation on how to build, and how to run each type of test is really helpful.

radical committed:
address review feedback from @ joperezr

radical committed:
address review feedback from @ joperezr

joperezr commented on tests/Aspire.Workload.Tests/WorkloadTestsBase.cs:
```diff
@@ -0,0 +1,252 @@
+                options = new BrowserTypeLaunchOptions
+                {
+                    Headless = false,
+                    ExecutablePath = "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"
```
NIT: should this be configurable (via env var or something) or is it always guaranteed to be there?

radical committed:
Add BROWSER_PATH to force a browser path to be used

eerhardt commented on Directory.Packages.props:
```diff
@@ -159,5 +159,6 @@
     <!-- Pinned version for Component Governance - Remove when root dependencies are updated -->
     <PackageVersion Include="Newtonsoft.Json" Version="13.0.3" />
     <PackageVersion Include="Newtonsoft.Json.Schema" Version="3.0.16" />
+    <PackageVersion Include="Microsoft.Playwright" Version="1.42.0" />
```
Can you move this to be under "unit test dependencies"?

eerhardt commented on tests/PlaywrightWrapper/Program.cs:
```diff
@@ -0,0 +1,8 @@
+sealed class Program
+{
+    public static int Main(string[] args)
+        => Microsoft.Playwright.Program.Main(args);
```
Why do we need this wrapper? Why can't we just call Microsoft.Playwright directly?

(It would be good to put the answer in a comment in the code)

radical replied:
Looks like creating a simple runtimeconfig.json, and deploying it next to the library is enough to run that directly:

```json
{
  "runtimeOptions": {
    "tfm": "net8.0",
    "framework": {
      "name": "Microsoft.NETCore.App",
      "version": "8.0.0"
    },
    "configProperties": {
      "System.Runtime.Serialization.EnableUnsafeBinaryFormatterSerialization": false
    }
  }
}
```

radical replied:
Dropped the project, and directly running the assembly now. It works locally.

eerhardt commented on tests/helix/send-to-helix-endtoendtests.targets:
```diff
@@ -32,11 +32,16 @@
-        <PreCommands Condition="'$(OS)' == 'Windows_NT'">set &quot;TEST_NAME=$(_WorkItemName)&quot; &amp; set TEST_SCENARIO=%(Identity) &amp; set &quot;CODE_COV_FILE_SUFFIX=-%(Identity)&quot;</PreCommands>
-        <PreCommands Condition="'$(OS)' != 'Windows_NT'">export &quot;TEST_NAME=$(_WorkItemName)&quot; &amp;&amp; export TEST_SCENARIO=%(Identity) &amp;&amp; export &quot;CODE_COV_FILE_SUFFIX=-%(Identity)&quot;</PreCommands>
+
+        <PreCommands                                         >$(_EnvVarSetKeyword) &quot;TEST_NAME=$(_WorkItemName)&quot;</PreCommands>
```
This doesn't look readable/maintainable. I would remove the whitespace.

eerhardt commented on tests/Aspire.EndToEnd.Tests/Directory.Build.props:
```diff
@@ -1,7 +1,12 @@
   <Import Project="..\Directory.Build.props" />
+
+  <PropertyGroup>
+    <RunSettingsFilePath>$(MSBuildThisFileDirectory).runsettings</RunSettingsFilePath>
```
Isn't this already set in the .csproj?

radical replied:
oops, yes!

eerhardt commented on tests/Shared/WorkloadTesting/AspireProject.cs:
```diff
@@ -36,13 +47,69 @@ public AspireProject(string id, string baseDir, ITestOutputHelper testOutput, Bu
             throw new InvalidOperationException("Project is already running");
         }
 
+        Stopwatch runTimeStopwatch = new();
```
Can you move this closer to where it is used?

eerhardt commented on tests/Shared/WorkloadTesting/AspireProject.cs:
```diff
@@ -199,6 +286,51 @@ public async Task BuildAsync(string[]? extraBuildArgs = default, CancellationTok
+        return dashboardPage;
+    }
+
+    public Task WaitForDashboardToBeAvailableAsync(CancellationToken token = default)
```
```suggestion
    public Task WaitForDashboardToBeAvailableAsync(CancellationToken cancellationToken = default)
```

This is a better name than `token`.

eerhardt commented on tests/Aspire.Workload.Tests/data/Endpoint_cs:
```diff
@@ -0,0 +1,42 @@
+using Aspire.Hosting.Lifecycle;
```
Can you rename this file to match the class name?

radical committed:
Merge remote-tracking branch 'origin/main' into workload-tests-prep

radical committed:
address review feedback from @ eerhardt

eerhardt commented on tests/Aspire.Workload.Tests/Aspire.Workload.Tests.csproj:
```diff
@@ -0,0 +1,59 @@
+
+    <RunTestsOnHelix>true</RunTestsOnHelix>
+    <TestUsingWorkloads>true</TestUsingWorkloads>
+    <InstallWorkloadForTesting>$(TestUsingWorkloads)</InstallWorkloadForTesting>
```
This is kind of weird. Should this just be the default if `TestusingWorkloads` is true?

radical replied:
Sure. It just shows the dependence between the two properties. In EndToEnd.Tests `TestUsingWorkloads` is `false` on CI/windows.

radical committed:
address review feedback from @ eerhardt, and cleanup some unnecessary code

eerhardt commented on tests/Aspire.Workload.Tests/Aspire.Workload.Tests.csproj:
```diff
@@ -0,0 +1,59 @@
+    <TestUsingWorkloads>true</TestUsingWorkloads>
+    <InstallWorkloadForTesting>$(TestUsingWorkloads)</InstallWorkloadForTesting>
+
+    <BuiltNuGetsDir>$(ArtifactsShippingPackagesDir)</BuiltNuGetsDir>
```
Why does this need to be set in each .csproj? Can't it be the default?

eerhardt commented on tests/Aspire.Workload.Tests/Aspire.Workload.Tests.csproj:
```diff
@@ -0,0 +1,59 @@
+    <InstallWorkloadForTesting>$(TestUsingWorkloads)</InstallWorkloadForTesting>
+
+    <BuiltNuGetsDir>$(ArtifactsShippingPackagesDir)</BuiltNuGetsDir>
+    <PackageVersionForWorkloadManifests>$(PackageVersion)</PackageVersionForWorkloadManifests>
```
Same for this - should be the default and not needed in the .csproj.

radical replied:
I could move these two to `tests/Directory.Build.props`.

eerhardt replied:
Why not to Workloads.Shared.targets? Isn't that what uses them?

radical committed:
Update tests/Shared/WorkloadTesting/AspireProject.cs

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

eerhardt commented on tests/Aspire.Workload.Tests/WorkloadTestsBase.cs:
```diff
@@ -0,0 +1,263 @@
+        }
+    }
+
+    internal static async Task AssertStarterTemplateRunAsync(IBrowserContext context, AspireProject project, string config, ITestOutputHelper _testOutput)
```
This is a bit odd to have in the WorkloadTestBase class. It seems like it is specific to certain tests.

eerhardt commented on tests/Aspire.Workload.Tests/WorkloadTestsBase.cs:
```diff
@@ -0,0 +1,263 @@
+        HashSet<string> foundNames = [];
+        List<ResourceRow> foundRows = [];
+
+        var timeout = TimeSpan.FromSeconds(timeoutSecs);
```
Can we use a CancellationToken here?

eerhardt commented on tests/Aspire.Workload.Tests/WorkloadTestsBase.cs:
```diff
@@ -0,0 +1,263 @@
+            return await playwright.Chromium.LaunchAsync(options).ConfigureAwait(false);
+        });
+
+        // default timeout for LaunchAsync is 30secs, so pick something larger than that
```
comment seems off from the code

radical replied:
Updated the comment. Please lemme know if it still seems off.

eerhardt commented on tests/Aspire.Workload.Tests/WorkloadTestsBase.cs:
```diff
@@ -0,0 +1,263 @@
+                        .Select(async e => await e.InnerTextAsync())
+                        .Select(t => t.Result.Trim(','))
+                        .ToArray();
+                // FIXME: this could still return "+2"
```
What's the plan to address this?

radical replied:
um I don't have any immediate plan on fixing this. I haven't hit this yet, but it is a possibility. And the comment might help somebody 🤷

eerhardt commented on tests/Aspire.Workload.Tests/WorkloadTestsBase.cs:
```diff
@@ -0,0 +1,263 @@
+
+    public static string FixupSymbolName(string name)
+    {
+        UTF8Encoding utf8 = new();
```
We shouldn't need to encode the name to UTF8 just to do this fix up / normalizing.

radical committed:
address review feedback from @ eerhardt

eerhardt commented on tests/Aspire.Workload.Tests/WorkloadTestsBase.cs:
```diff
@@ -0,0 +1,263 @@
+
+    // Don't fixup the prefix so it can have characters meant for testing, like spaces
+    public static string GetNewProjectId(string? prefix = null)
+        => (prefix is null ? "" : $"{prefix}_") + FixupSymbolName(Path.GetRandomFileName());
```
GetRandomFileName returns an 8.3 file name:

[https://github.com/dotnet/runtime/blob/0b10f4ba0a9174937829ec529c197dd42a569e36/src/libraries/System.Private.CoreLib/src/System/IO/Path.cs#L275-L279](https://github.com/dotnet/runtime/blob/0b10f4ba0a9174937829ec529c197dd42a569e36/src/libraries/System.Private.CoreLib/src/System/IO/Path.cs#L275-L279)

We should only need to replace the `.` (but even then - why? `.` are fine in project names)

radical replied:
This is originally from https://github.com/dotnet/runtime/commit/26990158d9725aa55d0af81f050a7a8e90c35f58 . But I changed this up to not fixup the whole name, but didn't get to deciding whether it could be dropped completely.

I think you are correct, and we can drop the fixup as our template generators are now to handle anything odd.

radical replied:
I tested this with some more tests I have for project names, and they passed with the fixup removed.

radical committed:
fix build

radical committed:
address review feedback from @ eerhardt - move common props to tests/Directory.Build.props

eerhardt commented on tests/Aspire.Workload.Tests/StarterTemplateRunTestsBase.cs:
```diff
@@ -0,0 +1,143 @@
+                Type: "Project",
+                Name: "apiservice",
+                State: "Running",
+                Source: $"{project!.Id}.ApiService.csproj",
```
Why the `!`?

radical replied:
good catch!

radical committed:
Move AssertStarterTemplateRunAsync to TemplateTests - address review feedback from @ eerhardt

radical committed:
address review feedback from @ eerhardt

radical committed:
address review feedback from @ eerhardt

radical committed:
Drop FixupSymbolName as it is not needed

The template generators can handle odd characters in the project name,
so this isn't needed anymore.

radical committed:
run playwright tests in headless chromium

radical committed:
Remove PlaywrightWrapper project

.. and instead use a pre-generated `runtimeconfig.json` to directly run
`Microsoft.Playwright.dll`.

Based on feedback from @ eerhardt .

radical committed:
dogfood.sh: accept a parameter for config. based on feedback from @ joperezr

radical committed:
Merge branch 'dotnet:main' into workload-tests-prep

eerhardt commented on tests/Aspire.Workload.Tests/data/nuget8.config:
```diff
@@ -0,0 +1,17 @@
+  </fallbackPackageFolders>
+  <packageSources>
+    <clear />
+    <!-- TEST_RESTORE_SOURCES_INSERTION_LINE -->
```
Is this necessary? What's it for?

radical replied:
This was used for inserting a source for the locally built nugets. But I changed the way we do that, so this line can be dropped.

radical replied:
Dropped.

eerhardt commented on tests/Aspire.Workload.Tests/data/nuget8.config:
```diff
@@ -0,0 +1,17 @@
+    <add key="dotnet8" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet8/nuget/v3/index.json" />
+    <add key="nuget.org"  value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
+  </packageSources>
+    <disabledPackageSources>
+    <clear />
+  </disabledPackageSources>
```
```suggestion
  <disabledPackageSources>
    <clear />
  </disabledPackageSources>
```

eerhardt commented on tests/Aspire.Workload.Tests/Aspire.Workload.Tests.csproj:
```diff
@@ -0,0 +1,56 @@
+
+    <RunTestsOnHelix>true</RunTestsOnHelix>
+    <TestUsingWorkloads>true</TestUsingWorkloads>
+    <InstallWorkloadForTesting>true</InstallWorkloadForTesting>
```
```suggestion
```

I don't think this is needed, since it defaults to `TestUsingWorkloads`

radical replied:
It doesn't default to `TestUsingWorkloads`.

The workload testing targets install two sdks - `dotnet-latest`, and `dotnet-none`. The former has the manifest+workload installed, but that latter has only the manifest installed. It is useful for running tests that want to check the case the user does not have the workload installed.

The `TestUsingWorkloads` property essentially controls that. I'll open a PR in `dotnet/runtime` to rename this property so its purpose is clearer. And possibly clean this up a bit.

radical replied:
Tracking in https://github.com/dotnet/aspire/issues/4222 .

eerhardt commented on tests/Aspire.Workload.Tests/Aspire.Workload.Tests.csproj:
```diff
@@ -0,0 +1,56 @@
+          Condition="'$(ExtractTestClassNamesForHelix)' == 'true'"
+          BeforeTargets="ZipTestArchive">
+
+    <Exec Command="&quot;$(DotNetTool)&quot; test --no-build -c $(Configuration) -s $(RunSettingsFilePath) -t --nologo -v:q -p:VsTestUseMSBuildOutput=false" ConsoleToMSBuild="true">
```
```suggestion
    <Exec Command="&quot;$(DotNetTool)&quot; test --no-build -c $(Configuration) -s $(RunSettingsFilePath) --list-tests --nologo -v:q -p:VsTestUseMSBuildOutput=false" ConsoleToMSBuild="true">
```

Makes it a little more understandable/readable.

eerhardt commented on tests/Aspire.Workload.Tests/Aspire.Workload.Tests.csproj:
```diff
@@ -0,0 +1,56 @@
+    </Exec>
+
+    <PropertyGroup>
+      <_Regex>^\s*(Aspire.Workload.Tests[^\($]+)</_Regex>
```
What happens if a new test gets added that doesn't follow this naming? Does the test just get skipped and never executed?

radical replied:
Yeah :/ I'll open an issue to make this more robust.

radical replied:
Tracking in https://github.com/dotnet/aspire/issues/4222 .

eerhardt commented on tests/Shared/WorkloadTesting/TestUtils.cs:
```diff
@@ -0,0 +1,55 @@
+
+namespace Aspire.Workload.Tests;
+
+public static class TestExtensions
```
Can the file name match the class name?

eerhardt commented on tests/Shared/WorkloadTesting/RequiresDockerDiscoverer.cs:
```diff
@@ -0,0 +1,20 @@
+{
+    public IEnumerable<KeyValuePair<string, string>> GetTraits(IAttributeInfo traitAttribute)
+    {
+        if (!RequiresDockerTheoryAttribute.IsSupported)
```
(not necessarily for this PR) - it would be nice if we had a consolidated "RequiresDocker" story across all our tests. This feels pretty spread out - some in `Aspire.Components.Common.Tests`, some here, some in `Aspire.Hosting.Tests`.

radical replied:
Tracking in https://github.com/dotnet/aspire/issues/4222 .

radical committed:
Update tests/Aspire.Workload.Tests/Aspire.Workload.Tests.csproj

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

radical committed:
Update tests/Aspire.Workload.Tests/data/nuget8.config

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

radical committed:
Merge remote-tracking branch 'origin/main' into workload-tests-prep

radical committed:
address review feedback from @ eerhardt

radical committed:
address review feedback from @ eerhardt

This was used for inserting a source for the locally built nugets. But this was changed to instead have `<add key="built-local" value="%BUILT_NUGETS_PATH%" />` so the path comes from the environment variable `BUILT_NUGETS_PATH` now, and we can avoid needing to patch the nuget.config.

radical committed:
nuget8.config: add dotnet-public, and remove nuget.org - feedback from @ eerhardt

eerhardt reviewed (approved):
Nice work here @radical! Let's get this in and we can iterate on it as we go.
